### PR TITLE
Use two vertex selectors in Jaccard and Dice

### LINKED ISSUE #175: Jaccard and Dice similarity functions should accept a from_vs and a to_vs instead of just a single vs
This would allow one to calculate, say, the histogram of the similarity values of large graphs when the whole similarity matrix does not fit into memory but a single row of the matrix does.

---

Imported from Launchpad using lp2gh.
- date created: 2009-09-23T10:01:52Z
- owner: ntamas
- the launchpad url was https://bugs.launchpad.net/bugs/435155

### PR DESCRIPTION
**Pull request by GroteGnoom**
Fixes #175

Lots of things still missing, like docs and functions.yaml, but I just want to check if this is what's expected.

### CODE REVIEW

**Comment by ntamas** [comment]
Yes, this looks good to me!

**Comment by GroteGnoom** [comment]
was 'DEPS: vids ON graph res, mode ON vids' for similarity_jaccard correct? Because it seems strange to me. I don't think `mode` should depend on `vids`, and `vids` shouldn't depend on `res`.

**Comment by ntamas** [comment]
I also think that ‘vids ON graph’ should be enough. The code generator needs the graph to convert the vertex IDs into an igraph_vs_t, but the mode is not necessary there.

**Review by ntamas — APPROVED** [review]
Ready to merge?

File: examples/simple/igraph_similarity.c [hunk 1]
```diff
@@ -49,7 +49,7 @@ int main(void) {
     }
 
     printf("Jaccard similarity:\n");
-    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), IGRAPH_ALL, 0);
+    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), igraph_vss_range(1,3), IGRAPH_ALL, 0);
```
Comment (szhorvat) [inline examples/simple/igraph_similarity.c:52]:
```suggestion
    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), igraph_vss_range(1, 3), IGRAPH_ALL, IGRAPH_NO_LOOPS);
```

File: examples/simple/igraph_similarity.c [hunk 2]
```diff
@@ -61,7 +61,7 @@ int main(void) {
     igraph_vector_print(&res);
 
     printf("\nDice similarity:\n");
-    igraph_similarity_dice(&g, &m, igraph_vss_range(1, 3), IGRAPH_ALL, 0);
+    igraph_similarity_dice(&g, &m, igraph_vss_range(1, 3), igraph_vss_range(1,3), IGRAPH_ALL, 0);
```
Comment (szhorvat) [inline examples/simple/igraph_similarity.c:64]:
```suggestion
    igraph_similarity_dice(&g, &m, igraph_vss_range(1, 3), igraph_vss_range(1, 3), IGRAPH_ALL, IGRAPH_NO_LOOPS);
```

File: src/misc/cocitation.c [hunk 3]
```diff
@@ -299,9 +299,11 @@ static igraph_error_t igraph_i_neisets_intersect(
...
  * \param res Pointer to a matrix, the result of the calculation will
  *        be stored here. The number of its rows and columns is the same
- *        as the number of vertex IDs in \p vids.
- * \param vids The vertex IDs of the vertices for which the
- *        calculation will be done.
+ *        as the number of vertex IDs in \p from and \p to, respectively.
+ * \param from The vertex IDs of the first vertices of the
+ *        pairs for which the calculation will be done.
+ * \param to The vertex IDs of the second vertices of the
+ *        pairs for which the calculation will be done.
```
Comment (szhorvat) [inline src/misc/cocitation.c:306]:
```suggestion
 * \param from The vertex IDs of the first set of vertices of the
 *        pairs for which the calculation will be done.
 * \param to The vertex IDs of the second set of vertices of the
 *        pairs for which the calculation will be done.
```

File: src/misc/cocitation.c [hunk 4]
```diff
@@ -335,23 +337,23 @@ static igraph_error_t igraph_i_neisets_intersect(
  * \example examples/simple/igraph_similarity.c
  */
 igraph_error_t igraph_similarity_jaccard(const igraph_t *graph, igraph_matrix_t *res,
-                              const igraph_vs_t vids, igraph_neimode_t mode, igraph_bool_t loops) {
+                              const igraph_vs_t from, const igraph_vs_t to, igraph_neimode_t mode, igraph_bool_t loops) {
     igraph_lazy_adjlist_t al;
     igraph_vit_t vit, vit2;
```
Comment (szhorvat) [inline src/misc/cocitation.c:342]: Can you rename these `vit_from` and `vit_to`? Should be a quick job for an IDE.

File: tests/unit/igraph_similarity.c [hunk 5]
```diff
@@ -154,7 +154,10 @@ int main(void) {
         return 1;
     }
 
-    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), IGRAPH_ALL, 0);
+    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), igraph_vss_range(1,3), IGRAPH_ALL, 0);
```
Comment (szhorvat) [inline tests/unit/igraph_similarity.c:157]:
```suggestion
    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), igraph_vss_range(1, 3), IGRAPH_ALL, IGRAPH_NO_LOOPS);
```

File: tests/unit/igraph_similarity.c [hunk 6]
```diff
@@ -154,7 +154,10 @@ int main(void) {
...
+    print_matrix(&m);
+
+    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), igraph_vss_1(1), IGRAPH_ALL, 0);
```
Comment (szhorvat) [inline tests/unit/igraph_similarity.c:160]:
```suggestion
    igraph_similarity_jaccard(&g, &m, igraph_vss_range(1, 3), igraph_vss_1(1), IGRAPH_ALL, IGRAPH_NO_LOOPS);
```

**Review by szhorvat — APPROVED** [review]
Only very minor stylistic nitpicks, otherwise LGTM.

**Comment by GroteGnoom** [comment]
Done.

**Comment by szhorvat** [comment]
I was unclear when talking about the renaming, I'll fix it on `master`

**Comment by GroteGnoom** [comment]
> I was unclear when talking about the renaming, I'll fix it on `develop`
> 
> EDIT: This is what I had in mind [f0c4c94](https://github.com/igraph/igraph/commit/f0c4c9453292ffe92957753f70b23931763d2f03)

Ha, of course! I mixed up iterators and selectors :grimacing:

**Pull request merged**